tests/copy_repo/test_copy_repo.py

```diff
@@ -0,0 +1,387 @@
+            assert isinstance(from_super, UdCacheClient)
+
+        # We'll substitute our own, only if UD client is being used
+        return self._udcache_client if from_super else None
+
+
+def test_missing_repos(command_tester):
+    """Command fails when pointed at nonexistent repos."""
+    # This one test covers the entry point
+
+    command_tester.test(
+        lambda: pubtools._pulp.tasks.copy_repo.entry_point(FakeCopyRepo),
+        [
+            "test-copy-repo",
+            "--pulp-url",
+            "https://pulp.example.com/",
+            "repo1,repo2 repo3,repo4",
+        ],
+    )
+
+
+def test_copy_empty_repo(command_tester, fake_collector):
+    """Copying a repo which is empty succeeds."""
+
+    with FakeCopyRepo() as task_instance:
+        repoA = FileRepository(id="some-filerepo")
+        repoB = FileRepository(id="another-filerepo")
+
+        task_instance.pulp_client_controller.insert_repository(repoA)
+        task_instance.pulp_client_controller.insert_repository(repoB)
+
+        command_tester.test(
+            task_instance.main,
+            [
+                "test-copy-repo",
+                "--pulp-url",
+                "https://pulp.example.com/",
+                "some-filerepo,another-filerepo",
+            ],
+        )
+
+    # No push items recorded
+    assert not fake_collector.items
+
+
+def test_copy_file_repo(command_tester, fake_collector):
+    """Copying a repo with file content succeeds."""
+
+    repoA = FileRepository(
+        id="some-filerepo",
+        eng_product_id=123,
+        relative_url="some/publish/url",
+        mutable_urls=["mutable1", "mutable2"],
+    )
+    repoB = FileRepository(
+        id="another-filerepo",
+        eng_product_id=456,
+        relative_url="another/publish/url",
+        mutable_urls=["mutable1", "mutable2"],
+    )
+
+    files = [
+        FileUnit(path="hello.txt", size=123, sha256sum="a" * 64),
+        FileUnit(path="with/subdir.json", size=0, sha256sum="b" * 64),
+    ]
+
+    with FakeCopyRepo() as task_instance:
+        fakepulp = task_instance.pulp_client_controller
+        fakepulp.insert_repository(repoA)
+        fakepulp.insert_repository(repoB)
+        # Populate source repository.
+        fakepulp.insert_units(repoA, files)
+
+        # It should run with expected output.
+        command_tester.test(
+            task_instance.main,
+            [
+                "test-copy-repo",
+                "--pulp-url",
+                "https://pulp.example.com/",
+                "--pulp-insecure",
+                "--udcache-url",
+                "https://ud.example.com/",
+                "some-filerepo,another-filerepo",
+            ],
+        )
+
+    # It should record that it copied these push items:
+    assert sorted(fake_collector.items, key=lambda pi: pi["filename"]) == [
+        {
+            "state": "PUSHED",
+            "origin": "pulp",
+            "src": None,
+            "dest": None,
```

Comment:
Why are the src and dest 'None' ?

Well, src I'm not entirely sure it's appropriate for it to be non-None.  But at least shouldn't `dest` be the destination repo of the copy?